tests/components/huawei_lte/test_config_flow.py

```diff
@@ -0,0 +1,148 @@
+    assert result["step_id"] == "user"
+
+
+async def test_urlize_plain_host(hass, requests_mock):
+    """Test that plain host or IP gets converted to a URL."""
+    requests_mock.request(ANY, ANY, exc=ConnectionError())
+    flow = ConfigFlowHandler()
+    flow.hass = hass
+    host = "192.168.100.1"
+    user_input = {**FIXTURE_USER_INPUT, CONF_URL: host}
+    result = await flow.async_step_user(user_input=user_input)
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    assert user_input[CONF_URL] == f"http://{host}/"
+
+
+async def test_already_configured(hass):
+    """Test we reject already configured devices."""
+    MockConfigEntry(
+        domain=DOMAIN, data=FIXTURE_USER_INPUT, title="Already configured"
+    ).add_to_hass(hass)
+
+    flow = ConfigFlowHandler()
+    flow.hass = hass
+    # Tweak URL a bit to check that doesn't fail duplicate detection
+    result = await flow.async_step_user(
+        user_input={
+            **FIXTURE_USER_INPUT,
+            CONF_URL: FIXTURE_USER_INPUT[CONF_URL].replace("http", "HTTP"),
+        }
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+
+async def test_connection_error(hass, requests_mock):
+    """Test we show user form on connection error."""
+
+    requests_mock.request(ANY, ANY, exc=ConnectionError())
+    flow = ConfigFlowHandler()
+    flow.hass = hass
+    result = await flow.async_step_user(user_input=FIXTURE_USER_INPUT)
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    assert result["errors"] == {CONF_URL: "connection_failed"}
+
+
+@pytest.fixture
+def login_requests_mock(requests_mock):
+    """Set up a requests_mock with base mocks for login tests."""
+    requests_mock.request(
+        ANY, FIXTURE_USER_INPUT[CONF_URL], text='<meta name="csrf_token" content="x"/>'
+    )
+    requests_mock.request(
+        ANY,
+        f"{FIXTURE_USER_INPUT[CONF_URL]}api/user/state-login",
+        text=(
+            f"<response><State>{LoginStateEnum.LOGGED_OUT}</State>"
+            f"<password_type>{PasswordTypeEnum.SHA256}</password_type></response>"
+        ),
+    )
+    return requests_mock
+
+
+async def _test_login_error(hass, req_mock, code, error_key, error_value):
+    req_mock.request(
+        ANY,
+        f"{FIXTURE_USER_INPUT[CONF_URL]}api/user/login",
+        text=f"<error><code>{code}</code><message/></error>",
+    )
+    flow = ConfigFlowHandler()
+    flow.hass = hass
+    result = await flow.async_step_user(user_input=FIXTURE_USER_INPUT)
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    assert result["errors"] == {error_key: error_value}
+
+
+async def test_incorrect_credentials(hass, login_requests_mock):
+    """Test we show user form on invalid credentials."""
+    await _test_login_error(
```

Comment:
It's possible to parametrize pytest tests:
https://docs.pytest.org/en/latest/parametrize.html#pytest-mark-parametrize

Reply:
Done.